Add a flag to control re-subscribe functionality.

redboltz opened the pull request:
This pull request implements a option `resubscribe` for controlling re-subscribe.
The default value of `resubscribe` is true. It doesn't break current behavior.
It will fix #664 partially. Partially means, if `suback` has error payload, then re-subscribe topics should be removed. It is a different issue.

mcollina commented:
Can you also add a fix for the other issue? There seem to be some new lines that are not covered by unit tests, would you mind adding those too?

Thanks!

redboltz commented:
> Can you also add a fix for the other issue? There seem to be some new lines that are not covered by unit tests, would you mind adding those too?

Does the other issue mean treat well with `suback` error? It's OK. I'm not sure the coverage will be satisfied if I add the fix. Anyway I will do.

In order to add the fix, I need to know how do I know `suback` error code.
http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718068
`suback`'s payload has two information. One is QoS, the other is success/fail bit.

And I found `packet.granted[0]` is the raw payload. Can I use this?
(packet.payload is null)

mcollina commented:
Yes it's the information you are looking for is in the granted property.

redboltz commented:
@mcollina , thanks. I implemented the library code. Now I start writing tests. I think that the test server needs to return `suback` with failure bit 1. How can I do this? I'm studying other tests implementation. But if you give me some advice, it is very helpful.

redboltz commented:
@mcollina , thank you! I will write tests.

redboltz commented:
I implemented the fix for suback error case. I also implemented the test.
It seems that codecov is satisfied.

I implemented the test in abstract_client.js. I'm not sure either abstract_client.js or client.js is an appropriate location to implement the test.
I introduced `port` variable and I set `server2` listening port to `port + 49`  just like the test you mentioned.

If there is a room to improve, please let me know.

mcollina commented:
Thanks, released as v2.12.0.
